CONTRIBUTING.md

```diff
@@ -1,47 +1,104 @@
+# Contributing to NGIpkgs
+
+This document is for people wanting to contribute to the implementation of NGIpkgs. This involves interacting with implementation changes that are proposed using [GitHub](https://github.com/) [pull requests](https://docs.github.com/pull-requests) to the [NGIpkgs](https://github.com/ngi-nix/ngipkgs/) repository (which you're in right now).
+
+This document assumes that you already know:
+
+- How to [create a Nix package](https://nix.dev/tutorials/packaging-existing-software#).
+- How to [use GitHub and Git](https://docs.github.com/en/get-started/quickstart/hello-world).
+
+In addition, a GitHub account is required, which you can sign up for [here](https://github.com/signup).
+
+## How to create pull requests
+
+Packagers are encouraged to contribute NGI projects to Nixpkgs, instead of to this repository. However, there are many reasons for not being able to do so, including:
+
+- Expediting the public availability of a package prior to its acceptance into Nixpkgs and landing in a channel.
+- Package is not a [good candidate for Nixpkgs](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#quick-start-to-adding-a-package).
+- Package is a good candidate for Nixpkgs, but no one is willing to be a maintainer :cry:.
+
+In any case, it is encouraged to create a pull request to Nixpkgs, then to this repository, with a comment linking to the pull request to Nixpkgs in the description and the Nix expressions.
+
+---
+
+Now that this is out of the way. To create pull requests to NGIpkgs:
+
+1. Set up a local version of NGIpkgs. If you don't have write access to NGIpkgs, create a [fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/fork-a-repo#forking-a-repository) of the repository and clone it.
+
+   ```ShellSession
+   git clone https://github.com/some-owner/ngipkgs.git
+   cd ngipkgs
+   ```
+
+1. Create and switch to a new Git branch.
+
+   ```ShellSession
+   git checkout -b some-branch
+   ```
+
+1. To add a package, start by creating a `package.nix` file in the package directory `pkgs/by-name/some-package`, where `some-package` will be the package attribute name.
+
+   ```ShellSession
+   mkdir -p pkgs/by-name/some-package
+   some-editor pkgs/by-name/some-package/package.nix
+   ```
+
+   Make sure to:
+
+   - Test the package on `x86_64-linux`.
+
+     ```ShellSession
+     git add pkgs/by-name/some-package
+     nix build .#some-package
+     ```
+
+   - Format the Nix expressions with [Alejandra](https://github.com/kamadorueda/alejandra).
+
+     ```ShellSession
+     alejandra pkgs/by-name/some-package
+     ```
+
+   To see how it’s done, an existing example is [libgnunetchat](https://github.com/ngi-nix/ngipkgs/blob/main/pkgs/by-name/libgnunetchat/package.nix).
+
+1. To add a module, start by creating a `default.nix` file in the module directory `projects/some-module`, where `some-module/service` will be the module name.
```

Comment:
I am not sure I understand this sentence correctly, but if I do, then it's on the verge of being outdated see #189.

Reply:
> Note that there's a special case, e.g.
> 
> https://github.com/ngi-nix/ngipkgs/blob/66dac53297cfc1273fe2f40baa7e8ed08f7f04c7/projects/Rosenpass/default.nix#L8
>
> where the path of the module is set to null. The semantics is that there are some options that are relevant to the project, but these are not provided by a module in NGIpkgs.

I don't understand what that means. Where are those options provided then?

Reply:
In a NixOS module that has been upstreamed to Nixpkgs. I am assuming that the same way that we want to re-export packages we want to re-export modules. We have #76 (which lacks a description) about this. The idea being that we do not want to capture strictly what's in NGIpkgs but also what was contributed to Nixpkgs.

Reply:
Oh, but why wouldn't we point that to the source location then? E.g.

```
nixos.modules.rosenpass = "${sources.nixpkgs}/nixos/modules/services/rosenpass.nix;
```

Reply:
Possible in principle, but we have to avoid loading it twice. If the module system figures this out since the path will be the same, it's easy and we have nothing to do. Otherwise we would have to figure out ourselves and that might be complicated. I can test that.

Reply:
It will figure out itself if the Nixpkgs instance for the module system evaluation is the same as for obtaining the module expression. We will want that documented very clearly because it may not be evident and fail in inscrutable ways if the condition is violated.